content/post/graceful-k8s-delpoyments/index.md

```diff
@@ -0,0 +1,205 @@
+
+```Dockerfile
+ROM alpine:3.5
+ 
+RUN apk add --no-cache bash
+
+#Install haproxy
+apk add haproxy=2.0.14-r0 
+
+# Download dumb-init
+ADD https://github.com/Yelp/dumb-init/releases/download/v1.2.0/dumb-init_1.2.0_amd64 /usr/local/bin/dumb-init
+RUN chmod +x /usr/local/bin/dumb-init
+ 
+#Run dumb-init with the signal rewrite definitons and haproxy as child process
+CMD ["/usr/local/bin/dumb-init","--rewrite","15:10","haproxy","-f","config.cfg"]
+```
+
+## Idea #1: Sleep in the pre-stop hook
+
+A common mitigation is to pause the pod eviction process in order to wait for the endpoint removal process to propagate throughout the Kubernetes cluster. After all, not only kube-proxies must be notified, but also other components such as ingress controllers and load balancers. To do that, we can modify use two settings in the [pod spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#podspec-v1-core): `spec.lifecycle.preStop` and `spec.terminationGracePeriodSeconds`.
+
+As described above, pre-stop hook executes before containers receive the `TERM` signal, so we can use it to achieve a graceful shutdown.
+To mitigate the problem, we add a `sleep` command in the pre-stop hook.
+This will suspend delay the `TERM` signal and create time for the endpoint removal to propagate.
+
+How long should we wait in the pre-stop hook?
+It depends on the latency of your network and the nodes.
+You may need to perform some tests to find out this value, however, multiple sources suggest a value between 5-10 seconds might be enough.
+
+Next, `spec.terminationGracePeriodSeconds` is the time limit that kubelet will wait before killing the container forcibly. The termination grace period starts from the execution of the pre-stop hook. Again there is no universally correct length for the grace period, you must calculate the value for `spec.terminationGracePeriodSeconds` based on your cluster and application behavior. It should include the sleep time in the pre-stop hook and the termination time of your application.
+
+For example a 20 second delay with up to 40 seconds of application shutdown time results in the following config:
+
+```yaml
+spec:
+  terminationGracePeriodSeconds: 60
+  containers:
+  - name: "{{APP_NAME}}"
+    lifecycle:
+      preStop:
+        exec:
+          command: ["/bin/sh","-c","sleep 20"]
+```
+
+An important note about long lived connections: the endpoint eviction does not affect established connections. So either you need a long enough `spec.terminationGracePeriodSeconds` to terminate long-lived connections or you need ensure that connections are short lived.
```

Comment:
Suggested:

```diff
- long lived connections
+ long lived TCP connections
```

QUIC connections rely on UDP and can be maintained across pods.